File: Sources/NotchleCore/Contracts/History.swift
import Foundation

// Play history (Tren, 2026-09-24: "keep the data of the ones i've gotten correct, in how many
// tries, etc. i want to be able to look back at it later in a diff tab within the notch thingy").
// The engine emits `GameEffect.recordOutcome` once per track; the platform stamps date and
// listing, appends a `HistoryEntry` and persists it with `HistoryStore`.

/// One played track.
public struct HistoryEntry: Sendable, Hashable, Codable, Identifiable {
    public let id: UUID
    public let date: Date
    public let trackID: String
    public let title: String
    public let artists: [String]
    public let listingName: String
    public let listingRef: SourceRef?
    public let correct: Bool
    /// Tier it was guessed at (0 = 5s); nil when missed.
    public let tierIndex: Int?
    public let wrongGuesses: Int
    public let skips: Int
    /// Album cover (Spotify oEmbed `thumbnail_url`, 300×300), resolved after the outcome.
    /// JSON key "artworkURL"; absent in older files (decodes as nil).
    public let artworkURL: URL?

    public init(id: UUID = UUID(), date: Date, trackID: String, title: String, artists: [String],
                listingName: String, listingRef: SourceRef?, correct: Bool, tierIndex: Int?,
                wrongGuesses: Int, skips: Int, artworkURL: URL? = nil) {
        self.id = id
        self.date = date
        self.trackID = trackID
        self.title = title
        self.artists = artists
        self.listingName = listingName
        self.listingRef = listingRef
        self.correct = correct
        self.tierIndex = correct ? tierIndex : nil
        self.wrongGuesses = wrongGuesses
        self.skips = skips
        self.artworkURL = artworkURL
    }

    /// The same entry with its cover set.
    public func withArtworkURL(_ url: URL?) -> HistoryEntry {
        HistoryEntry(id: id, date: date, trackID: trackID, title: title, artists: artists,
                     listingName: listingName, listingRef: listingRef, correct: correct,
                     tierIndex: tierIndex, wrongGuesses: wrongGuesses, skips: skips, artworkURL: url)
    }

    /// The entry for a `.recordOutcome` effect.
    public init(id: UUID = UUID(), date: Date, track: Track, outcome: TrackOutcome,
                wrongGuesses: Int, skips: Int, listingName: String, listingRef: SourceRef?,
                artworkURL: URL? = nil) {
        let tier: Int? = if case .correct(let t) = outcome { t } else { nil }
        self.init(id: id, date: date, trackID: track.id, title: track.title, artists: track.artists,
                  listingName: listingName, listingRef: listingRef, correct: tier != nil, tierIndex: tier,
                  wrongGuesses: wrongGuesses, skips: skips, artworkURL: artworkURL)
    }
}

/// Summary numbers over a history, pure.
public struct HistoryStats: Sendable, Hashable {
    public let total: Int
    public let correct: Int
    /// correct / total, 0 when empty.
    public let accuracy: Double
    /// Mean of (tierIndex + 1) over correct entries; nil when none is correct.
    public let averageTries: Double?
    /// Correct entries per tier index (index 0 = guessed at the first snippet). Sized to the
    /// highest tier seen.
    public let perTier: [Int]
    /// Longest run of consecutive correct entries, in date order.
    public let bestStreak: Int
    /// Run of correct entries ending at the newest entry.
    public let currentStreak: Int

    /// `entries` in any order; they are sorted by date (stable) first.
    public init(_ entries: [HistoryEntry]) {
        let ordered = entries.enumerated()
            .sorted { $0.element.date != $1.element.date ? $0.element.date < $1.element.date : $0.offset < $1.offset }
            .map(\.element)
        total = ordered.count
        let hits = ordered.filter(\.correct)
        correct = hits.count
        accuracy = total == 0 ? 0 : Double(correct) / Double(total)
        let tiers = hits.compactMap(\.tierIndex)
        averageTries = tiers.isEmpty ? nil : Double(tiers.reduce(0) { $0 + $1 + 1 }) / Double(tiers.count)
        var counts = Array(repeating: 0, count: (tiers.max() ?? -1) + 1)
        for t in tiers where t >= 0 { counts[t] += 1 }
        perTier = counts
        var best = 0, run = 0
        for e in ordered {
            run = e.correct ? run + 1 : 0
            best = max(best, run)
        }
        bestStreak = best
        currentStreak = run
    }
}

/// The spoiler rule: while a set is in progress its tracks may come round again (a replay after
/// a miss), so the history must not show them. They appear once the set ends
/// (`setComplete`/`setFailed`) or the player leaves the listing (idle, loading, exhausted).
public enum HistorySpoilerFilter {
    /// Track ids that must stay hidden in `state`.
    public static func hiddenTrackIDs(_ state: GameState) -> Set<String> {
        switch state.phase {
        case .idle, .loading, .setComplete, .setFailed, .exhausted:
            return []
        case .playingSnippet, .guessing, .wrong, .correct, .revealed, .error:
            return Set(state.currentSet.map(\.id))
        }
    }

    /// `entries` minus every entry for a hidden track, newest first.
    public static func visible(_ entries: [HistoryEntry], in state: GameState) -> [HistoryEntry] {
        let hidden = hiddenTrackIDs(state)
        return entries.filter { !hidden.contains($0.trackID) }.sorted { $0.date > $1.date }
    }
}

File: windows/src/Notchle.Core/Contracts/History.cs
namespace Notchle.Core;

// Mirror of Sources/NotchleCore/Contracts/History.swift: the play history (Tren, 2026-09-24:
// "keep the data of the ones i've gotten correct, in how many tries, etc. i want to be able to
// look back at it later in a diff tab within the notch thingy").

/// One played track, appended when its outcome is decided (GameEffect.RecordOutcome).
/// <param name="TierIndex">Tier the track was guessed at (0 = first try); null when missed.</param>
/// <param name="ListingRef">The playlist / album / artist it came from; null if unknown.</param>
/// <param name="ArtworkUrl">The album cover (ArtworkResolver), resolved after the outcome was
/// decided; null when unknown or offline.</param>
public sealed record HistoryEntry(
    Guid Id,
    DateTimeOffset Date,
    string TrackId,
    string Title,
    IReadOnlyList<string> Artists,
    string ListingName,
    SourceRef? ListingRef,
    bool Correct,
    int? TierIndex,
    int WrongGuesses,
    int Skips,
    Uri? ArtworkUrl = null)
{
    // Value equality over the artist list, like Track and the Swift struct.
    public bool Equals(HistoryEntry? other) =>
        other is not null && Id == other.Id && Date == other.Date && TrackId == other.TrackId
        && Title == other.Title && ListingName == other.ListingName && Equals(ListingRef, other.ListingRef)
        && Correct == other.Correct && TierIndex == other.TierIndex && WrongGuesses == other.WrongGuesses
        && Skips == other.Skips && Equals(ArtworkUrl, other.ArtworkUrl) && Artists.SequenceEqual(other.Artists);

    public override int GetHashCode() => HashCode.Combine(Id, Date, TrackId, Title, Correct, TierIndex, Artists.Count);
}

/// Totals over a history, oldest entry first. Pure.
/// <param name="AverageTries">Mean of TierIndex + 1 over the correct entries; null when none.</param>
/// <param name="CorrectByTier">Correct entries per tier: [0] = first try. As long as the highest
/// tier seen, empty when nothing was guessed.</param>
/// <param name="BestStreak">Longest run of consecutive correct entries.</param>
/// <param name="CurrentStreak">Run of correct entries at the end (the newest).</param>
public sealed record HistoryStats(
    int Total,
    int CorrectCount,
    double Accuracy,
    double? AverageTries,
    IReadOnlyList<int> CorrectByTier,
    int BestStreak,
    int CurrentStreak)
{
    public int MissedCount => Total - CorrectCount;

    public static HistoryStats From(IReadOnlyList<HistoryEntry> chronological)
    {
        var correct = chronological.Where(e => e.Correct && e.TierIndex is not null).ToList();
        var byTier = new int[correct.Count == 0 ? 0 : correct.Max(e => e.TierIndex!.Value) + 1];
        foreach (var e in correct) byTier[e.TierIndex!.Value]++;
        int best = 0, run = 0;
        foreach (var e in chronological)
        {
            run = e.Correct ? run + 1 : 0;
            best = Math.Max(best, run);
        }
        var total = chronological.Count;
        var correctCount = chronological.Count(e => e.Correct);
        return new HistoryStats(
            total,
            correctCount,
            total == 0 ? 0 : (double)correctCount / total,
            correct.Count == 0 ? null : correct.Average(e => e.TierIndex!.Value + 1.0),
            byTier,
            best,
            run);
    }

    // Value equality over the per-tier list.
    public bool Equals(HistoryStats? other) =>
        other is not null && Total == other.Total && CorrectCount == other.CorrectCount
        && Accuracy.Equals(other.Accuracy) && Nullable.Equals(AverageTries, other.AverageTries)
        && BestStreak == other.BestStreak && CurrentStreak == other.CurrentStreak
        && CorrectByTier.SequenceEqual(other.CorrectByTier);

    public override int GetHashCode() => HashCode.Combine(Total, CorrectCount, BestStreak, CurrentStreak);
}
